feature: SignatureValidation

esodot opened the pull request:
- Blockfrost webhook signature validation and verification functionality
- Refactoring and dependency updates

esodot committed:
feature: SignatureValidation

- Blockfrost webhook signature validation and verification functionality
- Refactoring and dependency updates

Commit:
feature: SignatureValidation

- Export the src/utils/blockfrost_signature_validator.dart

Commit:
feature: SignatureValidation

- Add own utils export file to control the exported files from utils folder.
- Use utils_exports.dart in main export file

Commit:
feature: SignatureValidation

- Add own utils export file to control the exported files from utils folder.
- Use utils_exports.dart in main export file

slowbackspace commented:
Hi @esodot, thanks for the contribution!
Would you be interested to add usage examples to our doscs on
https://blockfrost.dev/start-building/webhooks/webhooks-signatures#using-sdk
and
https://blockfrost.dev/start-building/webhooks/using-webhooks#process-a-webhook-request

(there is a "Edit this page" button at the bottom)

There is a bounty of 100 ADA for the task https://github.com/blockfrost/blockfrost.dev/issues/8 and you already did the hard part 🚀

andro-devs commented:
Hi @slowbackspace, thank you. I'm currently preparing the usage examples for the dart SDK and will create a pull request there ASAP.

andro-devs commented:
@slowbackspace  added PR into documentation repo: https://github.com/blockfrost/blockfrost.dev/pull/60

slowbackspace commented on lib/src/utils/blockfrost_signature_validator.dart:
```diff
@@ -0,0 +1,115 @@
+import 'package:blockfrost_api/src/utils/signature_validator.dart';
+import 'package:crypto/crypto.dart';
+
+const int maxToleranceSeconds = 60;
```
please make this configurable by using it as a fn param in `BlockfrostSignatureValidator.validate` with default of 600s. It should mimic TS/Python interface as closest as possible.

https://github.com/blockfrost/blockfrost-js/blob/master/src/utils/helpers.ts#L211

andro-devs replied:
fixed

slowbackspace commented on lib/src/utils/blockfrost_signature_validator.dart:
```diff
@@ -0,0 +1,115 @@
+/// Adapter class which implements the validator interface to validate the blockfrost webhook signature.
+class BlockfrostSignatureValidator implements SignatureValidator {
+  @override
+  bool validate({
```
Can you please change the order of the params to match the node.js, python and ruby SDK implementation

https://github.com/blockfrost/blockfrost-python/blob/master/blockfrost/helpers.py#L18
https://github.com/blockfrost/blockfrost-js/blob/master/src/utils/helpers.ts#L211

andro-devs replied:
fixed

andro-devs replied:
fixed

slowbackspace commented on lib/src/utils/blockfrost_signature_validator.dart:
```diff
@@ -0,0 +1,115 @@
+      providedSignatures.any((sig) => sig == expectedSignature);
+
+  if (!signatureMatch) {
+    print('Verification Failed: Signatures do not match.');
```
Leaking the expected HMAC = leaking an auth token for that payload+timestamp.
If anyone with log access sees the expected signature for the payload, they can replay the request within the tolerance window by sending the same body with t and v1 values from the logs.

If possible throw an error similiar to python/node.js implementation with signature header and payload attached. Don't print expected signature.
https://github.com/blockfrost/blockfrost-js/blob/master/src/utils/helpers.ts#L285-L291

andro-devs replied:
fixed

slowbackspace commented on lib/src/utils/blockfrost_signature_validator.dart:
```diff
@@ -0,0 +1,115 @@
+  print("timeDifference: $timeDifference");
+
+  if (timeDifference > maxToleranceSeconds) {
+    print(
```
same as above, please throw an error instead of using print fn

andro-devs replied:
fixed

slowbackspace reviewed (changes requested):
Good job! Just a few changes required. 
In general:
- The API for the verification function should stay consistent across all SDKs (order of fn parameters, same amount of configurability, same defaults, errors on failure)
- Instead of returning false, throw a custom error/exception in every failure case.

And don't forget to send your ADA address to receive the bounty once the PR is merged. Sadly there is a bounty is for the both sdk implementation and documentation so we can either send it to one address or split it between two.

Commit:
feature: Introduce SignatureValidationException

- Refactor print messages.
- Add SignatureValidationException.
- Use signature_validation_exception.dart in main export file.

andro-devs commented:
@slowbackspace Thank you for your review. I did the needed changes. Please review again. Also updated the PR in blockfrost.dev repo: https://github.com/blockfrost/blockfrost.dev/pull/60

Regarding Bounty, please use this address: 
addr1q9y86rykw0jdt0esf8g2y7y9dnp7hyp7tpwf4ez86jfzs4luw7nfzwm79hj9va6tek7nl5k2h4nu8urfvvc4wj0j5d2qq7u7ld
Thank you.

andro-devs commented:
Hi @slowbackspace any updates?

slowbackspace reviewed (approved):
Sorry for the delay. LGTM! The 100 ADA bounty will be paid shortly.

andro-devs commented:
if its simpler for you use the ada handle: "adam.app"

andro-devs commented:
> Sorry for the delay. LGTM! The 100 ADA bounty will be paid shortly.

Any updates here?

andro-devs commented:
> Sorry for the delay. LGTM! The 100 ADA bounty will be paid shortly.

@slowbackspace are you going to send it or not?

slowbackspace commented:
@andro-devs sorry, it's been super busy and the payment got stuck with the accounting department for a bit. we’ve sent something extra to make up for the delay.

tx fa9ad51ace3096c65cd97bc30533ea927d3f9e55dff0a806e764f3f0c9701746
